mq/relay.go

```diff
@@ -0,0 +1,298 @@
+func pubRelayedUpdate(client, relay *models.Client, peers *[]models.Client) {
+	//verify
+	if !logic.StringSliceContains(relay.Node.RelayedNodes, client.Node.ID.String()) {
+		logger.Log(0, "invalid call to pubRelayed update", client.Host.Name, relay.Host.Name)
+		return
+	}
+	//remove all nodes except relay
+	p := models.PeerAction{
+		Action: models.RemovePeer,
+	}
+	for _, peer := range *peers {
+		if peer.Host.ID == relay.Host.ID || peer.Host.ID == client.Host.ID {
+			continue
+		}
+		update := wgtypes.PeerConfig{
+			PublicKey: peer.Host.PublicKey,
+			Remove:    true,
+		}
+		p.Peers = append(p.Peers, update)
+	}
+	data, err := json.Marshal(p)
+	if err != nil {
+		logger.Log(0, "marshal peer update", err.Error())
+		return
+	}
+	fmt.Println("publishing peer update", client.Host.Name, p.Action, len(data))
+	publish(&client.Host, fmt.Sprintf("peer/host/%s/%s", client.Host.ID.String(), servercfg.GetServer()), data)
+	//update the relay peer
+	p = models.PeerAction{
+		Action: models.UpdatePeer,
+	}
+	update := wgtypes.PeerConfig{
+		PublicKey:         relay.Host.PublicKey,
+		ReplaceAllowedIPs: true,
+		Endpoint: &net.UDPAddr{
+			IP:   relay.Host.EndpointIP,
+			Port: relay.Host.ListenPort,
+		},
+		PersistentKeepaliveInterval: &relay.Node.PersistentKeepalive,
+	}
+	if relay.Node.Address.IP != nil {
+		relay.Node.Address.Mask = net.CIDRMask(32, 32)
+		update.AllowedIPs = append(update.AllowedIPs, relay.Node.Address)
+	}
+	if relay.Node.Address6.IP != nil {
+		relay.Node.Address6.Mask = net.CIDRMask(128, 128)
+		update.AllowedIPs = append(update.AllowedIPs, relay.Node.Address6)
+	}
+	p.Peers = append(p.Peers, update)
+	// add all other peers to allowed ips
+	for _, peer := range *peers {
+		if peer.Host.ID == relay.Host.ID || peer.Host.ID == client.Host.ID {
+			continue
+		}
+		if peer.Node.Address.IP != nil {
+			peer.Node.Address.Mask = net.CIDRMask(32, 32)
+			update.AllowedIPs = append(update.AllowedIPs, peer.Node.Address)
+		}
+		if peer.Node.Address6.IP != nil {
+			peer.Node.Address6.Mask = net.CIDRMask(128, 128)
+			update.AllowedIPs = append(update.AllowedIPs, peer.Node.Address6)
+		}
+		if peer.Node.IsRelay {
+			update.AllowedIPs = append(update.AllowedIPs, getAllowedIPs(peer)...)
+		}
+		if peer.Node.IsEgressGateway {
+			update.AllowedIPs = append(update.AllowedIPs, getEgressIPs(peer)...)
+		}
+		if peer.Node.IsIngressGateway {
+			update.AllowedIPs = append(update.AllowedIPs, getIngressIPs(peer)...)
+		}
+	}
+	p.Peers = append(p.Peers, update)
+	data, err = json.Marshal(p)
+	if err != nil {
+		logger.Log(0, "marshal peer update", err.Error())
+		return
+	}
+	fmt.Println("publishing peer update", client.Host.Name, p.Action, len(data))
+	publish(&client.Host, fmt.Sprintf("peer/host/%s/%s", client.Host.ID.String(), servercfg.GetServer()), data)
+}
+
+// pubRelayUpdate - publish peer update to a relay
+func pubRelayUpdate(client *models.Client, peers *[]models.Client) {
```

Comment:
Pointer to a slice? Whats the reason?

Reply:
consistency so all parameter are passed by ref

Reply:
Slices and maps are always by ref in golang (besides metadata). Why a double pointer? It introduces unnecessary complexity, but maybe theres a benefit which is not clear from the code?